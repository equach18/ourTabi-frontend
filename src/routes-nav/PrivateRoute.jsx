import { Navigate, Outlet } from "react-router-dom";

/** Protects routes that require authentication */
function PrivateRoute({ currentUser, isLoading }) {

  if (isLoading) return <Spinner />;

  return currentUser ? <Outlet /> : <Navigate to="/login" />;
}

export default PrivateRoute;
