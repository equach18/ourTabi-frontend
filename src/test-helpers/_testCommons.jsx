import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
