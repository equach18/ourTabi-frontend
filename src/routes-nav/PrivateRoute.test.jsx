import { screen } from "@testing-library/react";
import { Route, Routes } from "react-router-dom";
import PrivateRoute from "./PrivateRoute";
import { renderWithAllProviders } from "../test-helpers/_testCommons";

// mock Spinner
vi.mock("../components/common/Spinner", () => ({
  default: () => <div>Loading Spinner...</div>,
}));

describe("PrivateRoute", () => {
  function renderPrivateRoute({
    currentUser = null,
    isLoading = false,
    route = "/",
  } = {}) {
    return renderWithAllProviders(
      <Routes>
        <Route
          element={
            <PrivateRoute currentUser={currentUser} isLoading={isLoading} />
          }
        >
          <Route path="/" element={<div>Protected Content</div>} />
        </Route>
        <Route path="/login" element={<div>Login Page</div>} />
      </Routes>,
      {
        route,
        userContext: { currentUser },
      }
    );
  }
  
  it("matches snapshot when loading", () => {
    const { asFragment } = renderPrivateRoute({ isLoading: true });
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when authenticated", () => {
    const { asFragment } = renderPrivateRoute({
      currentUser: { username: "elaine" },
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when unauthenticated", () => {
    const { asFragment } = renderPrivateRoute();
    expect(asFragment()).toMatchSnapshot();
  });
  it("shows spinner when loading", () => {
    renderPrivateRoute({ isLoading: true });
    expect(screen.getByText("Loading Spinner...")).toBeInTheDocument();
  });

  it("renders protected content when user is logged in", () => {
    renderPrivateRoute({ currentUser: { username: "elaine" } });
    expect(screen.getByText("Protected Content")).toBeInTheDocument();
  });

  it("redirects to login when user is not logged in", () => {
    renderPrivateRoute();
    expect(screen.getByText("Login Page")).toBeInTheDocument();
  });
});
