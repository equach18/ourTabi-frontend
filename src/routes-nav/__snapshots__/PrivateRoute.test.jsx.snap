// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrivateRoute > matches snapshot when authenticated 1`] = `
<DocumentFragment>
  <div>
    Protected Content
  </div>
</DocumentFragment>
`;

exports[`PrivateRoute > matches snapshot when loading 1`] = `
<DocumentFragment>
  <div>
    Loading Spinner...
  </div>
</DocumentFragment>
`;

exports[`PrivateRoute > matches snapshot when unauthenticated 1`] = `
<DocumentFragment>
  <div>
    Login Page
  </div>
</DocumentFragment>
`;
